### PR TITLE
system refactorings

### PR DESCRIPTION
**Pull request by Araq**

### CODE REVIEW

File: lib/system/widestrs.nim [hunk 1]
```diff
@@ -10,8 +10,8 @@
 # Nim support for C/C++'s `wide strings`:idx:. This is part of the system
 # module! Do not import it directly!
```
Comment (timotheecour) [inline lib/system/widestrs.nim]: > This is part of the system ...
> Do not import it directly!

can be removed

File: lib/system/threads.nim [hunk 2]
```diff
@@ -325,11 +325,8 @@ when not defined(useNimRtl):
 
   when emulatedThreadVars:
     if nimThreadVarsSize() > sizeof(ThreadLocalStorage):
-      echo "too large thread local storage size requested ",
-           "(", nimThreadVarsSize(), "/", sizeof(ThreadLocalStorage), "). ",
-           "Use -d:\"nimTlsSize=", nimThreadVarsSize(),
-           "\" to preallocate sufficient storage."
-
+      c_fprintf(cstderr, """too large thread local storage size requested,
+use -d:\"nimTlsSize=X\" to setup even more or stop using unittest.nim""")
```
Comment (timotheecour) [inline lib/system/threads.nim]: > or stop using unittest.nim

huh? TLS usage can come from any other module not just unittest, that comment seems off

File: lib/system/sysio.nim [hunk 3]
```diff
@@ -15,439 +15,5 @@
 {.push debugger:off .} # the user does not want to trace a part
```
Comment (timotheecour) [inline lib/system/sysio.nim]: why not remove all code from here and put: `{.deprecated: "use `import io`".}`

File: lib/system/ansi_c.nim [hunk 4]
```diff
@@ -111,22 +111,27 @@ type c_sighandler_t = proc (a: cint) {.noconv.}
 proc c_signal(sign: cint, handler: proc (a: cint) {.noconv.}): c_sighandler_t {.
   importc: "signal", header: "<signal.h>", discardable.}
 
-proc c_fprintf(f: File, frmt: cstring): cint {.
+type
+  CFile {.importc: "FILE", header: "<stdio.h>",
+          incompletestruct.} = object
+  CFileStar* = ptr CFile ## The type representing a file handle.
```
Comment (timotheecour) [inline lib/system/ansi_c.nim]: why not `s/CFileStar/CFilePtr` ? seems more readable and it follows the existing nim convention

File: lib/system/endb.nim [hunk 5]
```diff
@@ -76,29 +76,52 @@ proc `==`(a, b: StaticStr): bool =
...
+proc write(f: CFileStar, s: StaticStr) =
   write(f, cstring(unsafeAddr s.data))
 
+proc write(f: CFileStar, i: int) =
+  when sizeof(int) == 8:
+    discard c_fprintf(f, "%lld", i)
+  else:
+    discard c_fprintf(f, "%ld", i)
+
+proc close(f: CFileStar): cint {.
```
Comment (timotheecour) [inline lib/system/endb.nim]: - [ ] why not follow existing convention, ie `c_fclose` (which was previous name in sysio.nim)
- [ ] like you said yourself, `discardable` is evil; it also makes errors silently accepted

for reference here was the previous definition:
```nim
proc c_fclose(f: File): cint {.
  importc: "fclose", header: "<stdio.h>".}
```

Comment (Araq) [reply lib/system/endb.nim]: it's ENDB, I hacked it together.

File: lib/system.nim [hunk 6]
```diff
@@ -3153,257 +3153,45 @@ when not defined(JS): #and not defined(nimscript):
...
+    proc moveMem(dest, source: pointer, size: Natural) =
+      c_memmove(dest, source, size)
+      when declared(memTrackerOp):
+        memTrackerOp("moveMem", dest, size)
+    proc equalMem(a, b: pointer, size: Natural): bool =
+      nimCmpMem(a, b, size) == 0
 
   proc cmp(x, y: string): int =
-    when nimvm:
+    when defined(nimscript):
```
[the remark this answers is not in the document]
Comment (Araq) [reply lib/system.nim]: That wouldn't compile. :P

File: lib/system/ansi_c.nim (hunk 4 above)
Comment (Araq) [reply lib/system/ansi_c.nim]: You're right, will change it.

File: lib/system/sysio.nim (hunk 3 above)
Comment (Araq) [reply lib/system/sysio.nim]: Agreed.

File: lib/system/threads.nim (hunk 2 above)
Comment (Araq) [reply lib/system/threads.nim]: That was hacked together. ;-)

File: lib/system/widestrs.nim (hunk 1 above)
Comment (Araq) [reply lib/system/widestrs.nim]: True.

**Pull request merged**